image.py

```diff
@@ -0,0 +1,355 @@
+import io
+from shlex import split
+from subprocess import CalledProcessError, run
+from sys import stdin
+from typing import Union
+
+from exceptions import CommandNotFoundOnImageError
+
+
+class Image:
+    """A class which holds a reference to a given docker image and provides
+    an interface by which to interact with that image.
+
+    -   Capable of building docker images from dockerfiles or dockerfile-
+        formatted strings via the build classmethods.
+    -   Capable of acquiring information about the given image (see property
+        methods below).
+    -   Capable of running commands on containers built from the image using
+        the run method.
+    -   All docker build and run methods on the class can recive a file-like
+        object, (io.TextIOBase or some subclass of it,) and they will write the
+        outputs from docker into the file to enable logging.
+
+    NOTE: This class does not work correctly if the image built does not have
+        bash installed.
+    """
+
+    def __init__(self, name_or_id: str):
+        """Initialize this Image. Connects the class with its ID.
+
+        Parameters
+        ----------
+        name_or_id : str
+            A name or ID by which to find this image using docker inspect.
+        """
+        self._id = get_image_id(name_or_id)
+
+    @classmethod
+    def build_from_dockerfile(
+        cls,
+        context: str,
+        tag: str,
+        output_file: Union[io.TextIOBase, None] = None,
+        dockerfile_loc: str = ""
```

Comment:
Another suggestion that's entirely up to personal preference, so feel free to ignore--

Sometimes it's nice to make some input parameters [keyword-only](https://peps.python.org/pep-3102/). This can be useful as a way of distinguishing the "core" arguments from secondary options. It forces callers of the function to be explicit when specifying options by preventing them from being passed as positional arguments.

In general, I typically recommend trying to be conservative and not making too many keyword-only arguments. In this case, though, it feels like `context` is the only "core" argument and the rest feel like options that I'd be comfortable making keyword-only.

So you might consider inserting a `*, ` after the `context` parameter, like this:
```suggestion
        cls,
        context: str,
        *,
        tag: str,
        output_file: Union[io.TextIOBase, None] = None,
        dockerfile_loc: str = ""
```